File: app/src/main/java/com/google/android/apps/translate/TranslateActivity.java
package com.google.android.apps.translate;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;

public class TranslateActivity extends Activity {

    private static final String OFFLINE_PACKAGE = "dev.davidv.translator";
    private static final String OFFLINE_MAIN_ACTIVITY = ".MainActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d("GOTr", "TranslateActivity started");

        Intent incomingIntent = getIntent();
        if (incomingIntent != null) {
            redirectToOffline(incomingIntent);
        }
        finish();
    }

    private void redirectToOffline(Intent incomingIntent) {
        Log.d("GOTr", "Redirecting to offline translator");
        
        Intent offlineIntent = new Intent(incomingIntent);
        offlineIntent.setComponent(new ComponentName(OFFLINE_PACKAGE, OFFLINE_PACKAGE + OFFLINE_MAIN_ACTIVITY));

        String action = incomingIntent.getAction();
        if (Intent.ACTION_SEND.equals(action)) {
            if ("text/plain".equals(incomingIntent.getType())) {
                String sharedText = incomingIntent.getStringExtra(Intent.EXTRA_TEXT);
                if (!TextUtils.isEmpty(sharedText)) {
                    offlineIntent.putExtra(Intent.EXTRA_TEXT, sharedText);
                    Log.d("GOTr", "Sending text: " + sharedText);
                }
            } else if (incomingIntent.getType() != null && incomingIntent.getType().startsWith("image/")) {
                Uri imageUri = incomingIntent.getParcelableExtra(Intent.EXTRA_STREAM);
                if (imageUri != null) {
                    offlineIntent.putExtra(Intent.EXTRA_STREAM, imageUri);
                    Log.d("GOTr", "Sending image: " + imageUri);
                }
            }
        } else if ("android.intent.action.TRANSLATE".equals(action)) {
            String textToTranslate = incomingIntent.getStringExtra(Intent.EXTRA_TEXT);
            if (!TextUtils.isEmpty(textToTranslate)) {
                offlineIntent.putExtra(Intent.EXTRA_TEXT, textToTranslate);
                Log.d("GOTr", "Translating text: " + textToTranslate);
            }
        }

        startActivity(offlineIntent);
        Log.d("GOTr", "Successfully started offline translator");
    }
}

File: app/src/main/java/com/google/android/apps/translate/TranslateActivity.kt
package com.google.android.apps.translate

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast

class TranslateActivity : Activity() {
    companion object {
        private const val TAG = "TranslateOverlay"
        private const val TARGET_PACKAGE = "dev.davidv.translator"
        private const val TARGET_ACTIVITY = ".ProcessTextActivity"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Redukcja migotania – od razu przekieruj
        handleIntent(intent)
        finish()
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        intent?.let { handleIntent(it) }
    }

    private fun handleIntent(intent: Intent) {
        val action = intent.action
        val text: String?
        val newIntent = Intent(Intent.ACTION_PROCESS_TEXT).apply {
            setComponent(ComponentName(TARGET_PACKAGE, TARGET_PACKAGE + TARGET_ACTIVITY))
            setType("text/plain")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)  // Bez migotania
        }

        when (action) {
            Intent.ACTION_PROCESS_TEXT -> {
                text = intent.getStringExtra(Intent.EXTRA_PROCESS_TEXT)
                val readonly = intent.getBooleanExtra(Intent.EXTRA_PROCESS_TEXT_READONLY, false)
                Log.d(TAG, "PROCESS_TEXT: $text, readonly=$readonly")
                newIntent.putExtra(Intent.EXTRA_PROCESS_TEXT, text)
                if (readonly) newIntent.putExtra(Intent.EXTRA_PROCESS_TEXT_READONLY, true)
            }
            Intent.ACTION_SEND -> {
                if ("text/plain" == intent.type) {
                    text = intent.getStringExtra(Intent.EXTRA_TEXT)
                    Log.d(TAG, "SEND: $text")
                    newIntent.putExtra(Intent.EXTRA_PROCESS_TEXT, text)
                } else return
            }
            else -> {
                text = intent.getStringExtra(Intent.EXTRA_PROCESS_TEXT) ?: intent.getStringExtra(Intent.EXTRA_TEXT)
                Log.d(TAG, "Generic: $text")
                newIntent.putExtra(Intent.EXTRA_PROCESS_TEXT, text)
            }
        }

        try {
            startActivity(newIntent)
            Log.d(TAG, "Redirected to $TARGET_PACKAGE")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to start $TARGET_PACKAGE: ${e.message}")
            Toast.makeText(this, getString(R.string.error_cannot_open_translator), Toast.LENGTH_SHORT).show()
        }
    }
}
